crates/chain/src/tx_graph.rs

```diff
@@ -232,176 +498,446 @@ impl TxGraph {
+    ///
+    /// Each transaction is represented as a [`CanonicalTx`] that contains where the transaction is
+    /// observed in-chain, and the [`TxNode`].
+    ///
+    /// # Error
+    ///
+    /// If the [`ChainOracle`] implementation (`chain`) fails, an error will be returned with the
+    /// returned item.
+    ///
+    /// If the [`ChainOracle`] is infallible, [`list_chain_txs`] can be used instead.
+    ///
+    /// [`list_chain_txs`]: Self::list_chain_txs
+    pub fn try_list_chain_txs<'a, C>(
+        &'a self,
+        chain: &'a C,
+        chain_tip: BlockId,
+    ) -> impl Iterator<Item = Result<CanonicalTx<'a, Transaction, A>, C::Error>>
+    where
+        C: ChainOracle + 'a,
+    {
+        self.full_txs().filter_map(move |tx| {
+            self.try_get_chain_position(chain, chain_tip, tx.txid)
+                .map(|v| {
+                    v.map(|observed_in| CanonicalTx {
+                        observed_as: observed_in,
+                        node: tx,
+                    })
+                })
+                .transpose()
         })
     }
 
-    /// Creates an iterator that filters and maps descendants from the starting `txid`.
+    /// List graph transactions that are in `chain` with `chain_tip`.
     ///
-    /// The supplied closure takes in two inputs `(depth, descendant_txid)`:
+    /// This is the infallible version of [`try_list_chain_txs`].
     ///
-    /// * `depth` is the distance between the starting `txid` and the `descendant_txid`. I.e., if the
-    ///     descendant is spending an output of the starting `txid`; the `depth` will be 1.
-    /// * `descendant_txid` is the descendant's txid which we are considering to walk.
+    /// [`try_list_chain_txs`]: Self::try_list_chain_txs
+    pub fn list_chain_txs<'a, C>(
+        &'a self,
+        chain: &'a C,
+        chain_tip: BlockId,
+    ) -> impl Iterator<Item = CanonicalTx<'a, Transaction, A>>
+    where
+        C: ChainOracle + 'a,
+    {
+        self.try_list_chain_txs(chain, chain_tip)
+            .map(|r| r.expect("oracle is infallible"))
+    }
+
+    /// List outputs that are in `chain` with `chain_tip`.
     ///
-    /// The supplied closure returns an `Option<T>`, allowing the caller to map each node it vists
-    /// and decide whether to visit descendants.
-    pub fn walk_descendants<'g, F, O>(&'g self, txid: Txid, walk_map: F) -> TxDescendants<F>
+    /// Floating ouputs are not iterated over.
+    ///
+    /// The `filter_predicate` should return true for outputs that we wish to iterate over.
+    ///
+    /// # Error
+    ///
+    /// A returned item can error if the [`ChainOracle`] implementation (`chain`) fails.
+    ///
+    /// If the [`ChainOracle`] is infallible, [`list_chain_txouts`] can be used instead.
+    ///
+    /// [`list_chain_txouts`]: Self::list_chain_txouts
+    pub fn try_list_chain_txouts<'a, C, P>(
+        &'a self,
+        chain: &'a C,
+        chain_tip: BlockId,
+        mut filter_predicate: P,
```

Comment:
Why can't I just filter the iterator that it returns?

Reply:
It's more efficient as we won't need to call `try_get_chain_spend` for every txout.

Reply:
You know what they say about premature optimization.... I'm strongly in favor of removing this.

Reply:
Removed.